src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java

```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
         assertSamValid(sam, ValidationStringency.LENIENT, null);
     }
 
+    /**
+     * Compares SAM/BAM files in a stringent way but not by byte identity (allow reorder of attributes)
+     */
+    private static String samsEqualLenient(File sam1, File sam2, ValidationStringency validation, File reference) throws IOException {
+        try(final SamReader reader1 = getReader(sam1, validation, reference);
+            final SamReader reader2 = getReader(sam2, validation, reference)) {
+
+            final SamComparison comparison = new SamComparison(reader1, reader2);
+            return comparison.areEqual() ? null : "SamComparison fails";
+        }
+    }
+
+    /**
+     * Compares SAM/BAM files in a stringent way but not by byte identity (allow reorder of attributes)
+     * Comparing by MD5s is too strict and comparing by SamComparison is too lenient. So we need this method.
+     * Returns null if equal or message string if not equal.
+     */
+    public static String samsEqualStringent(final File sam1, final File sam2, final ValidationStringency validation, final File reference) throws IOException {
+        //Three phases:
+        //1:  if MD5s are same, then we're equal. Quick check
+        final String fileMD5_1 = Utils.calculateFileMD5(sam1);
+        final String fileMD5_2 = Utils.calculateFileMD5(sam2);
+        if (fileMD5_1.equals(fileMD5_2)){
+            return null;
+        }
+
+        //2:  SamComparison says no, then we're not equal
+        try(final SamReader reader1 = getReader(sam1, validation, reference);
+            final SamReader reader2 = getReader(sam2, validation, reference)){
+
+            final SamComparison comparison = new SamComparison(reader1, reader2);
+            if (! comparison.areHeadersEqual()){  //For headers, we want identity (at least for now)
+                return "Headers different";
+            }
+            if (!comparison.areEqual()) {
+                return "SamComparison failed";
+            }
+        }
+
+        //At this point we know that the files are not byte-wise identical, but are equal according to SamComparison and their headers are equal
```

Comment:
Why both with the `SamComparison` step? Why not write a `compareHeadersIgnorePGAndCOLines()` method, call that, then go straight into the record-by-record `readsEqualAllowAddingAttributes()` comparison?